py/server/deephaven/table_factory.py

```diff
@@ -186,3 +194,77 @@ def write_row(self, *values: Any) -> None:
             self._j_table_writer.logRowPermissive(values)
         except Exception as e:
             raise DHError(e, "failed to write a row.") from e
+
+
+class InputTable(Table):
+    """InputTable is a subclass of Table that allows the users to dynamic add/delete/modify data in it. There are two
+    types of InputTable - append-only and keyed.
+
+    The append-only input table is not keyed, all rows are added to the end of the table, and deletions and edits are
+    not permitted.
+
+    The keyed input tablet has keys for each row and supports addition/deletion/modification of rows by the keys.
+    """
+
+    def __init__(self, col_defs: Dict[str, DType] = None, init_table: Table = None, key_cols: Union[str, Sequence[str]] = None):
+        """Creates an InputTable instance from either column definitions or initial table. When key columns are
+        provided, the InputTable will be keyed, otherwise it will be append-only.
+
+        Args:
+            col_defs (Dict[str, DType]): the column definitions
+            init_table (Table): the initial table
+            key_cols (Union[str, Sequence[str]): the name(s) of the key column(s)
+        """
+        try:
+            if col_defs is None and init_table is None:
+                raise ValueError("either column definitions or init table should be provided.")
+            elif col_defs and init_table:
+                raise ValueError("both column definitions and init table are provided.")
+
+            if col_defs:
+                cols = [Column(name=n, data_type=t) for n, t in col_defs.items()]
+                j_arg_1 = _JTableDefinition.of([col.j_column_definition for col in cols])
+            else:
+                j_arg_1 = init_table.j_table
+
+            key_cols = to_sequence(key_cols)
+            if key_cols:
+                super().__init__(_JKeyedArrayBackedMutableTable.make(j_arg_1, key_cols))
+            else:
+                super().__init__(_JAppendOnlyArrayBackedMutableTable.make(j_arg_1))
+            self.j_input_table = self.j_table.getAttribute(_J_INPUT_TABLE_ATTRIBUTE)
+            self.key_columns = key_cols
+        except Exception as e:
+            raise DHError(e, "failed to create a InputTable.") from e
+
+    def add(self, table: Table):
+        """Writes rows in the provided table to this input table. If this is keyed input table, added rows with keys
+        that match existing rows will replace those rows.
+
+        Args:
+            table (Table): the table that provides the rows to write
+
+        Raises:
+            DHError
+        """
+        try:
+            self.j_input_table.add(table.j_table)
+        except Exception as e:
+            raise DHError(e, "add to InputTable failed.") from e
+
+    def delete(self, table: Table):
+        """Deletes the keys contained in the provided table from this keyed input table.
```

Comment:
Doc string is written as if the input table is always a keyed input table, but it could be keyed or append.